src/storage/src/storage_state.rs

```diff
@@ -254,124 +349,208 @@ impl<'w, A: Allocate> Worker<'w, A> {
                 last_stats_time = Some(Instant::now());
             }
 
+            // Worker #0 pushes external commands into the sequencer. All other
+            // workers read them away but ignore them. Instead, other workers
+            // will get commands broadcast from the sequencer.
+            //
+            // NOTE: This seems hairy, but I don't yet see an alternative. We
+            // need a consistent order/interleaving of internal and external
+            // commands, across all workers, so we can't let each worker read
+            // their own external command stream because that would be a side
+            // channel and the ordering could (will) be arbitrary.
+
             // Handle any received commands.
             let mut cmds = vec![];
             let mut empty = false;
             while !empty {
                 match command_rx.try_recv() {
-                    Ok(cmd) => cmds.push(cmd),
+                    Ok(cmd) => {
+                        cmds.push(cmd);
+                    }
                     Err(TryRecvError::Empty) => empty = true,
                     Err(TryRecvError::Disconnected) => {
                         empty = true;
                         disconnected = true;
                     }
                 }
             }
-            for cmd in cmds {
-                self.handle_storage_command(cmd);
+            {
+                let mut internal_cmd_tx = command_sequencer.borrow_mut();
+                for cmd in cmds {
+                    self.storage_state.handle_storage_command(
+                        self.timely_worker.index(),
+                        &mut *internal_cmd_tx,
+                        &mut async_worker,
+                        cmd,
+                    )
+                }
+            }
+
+            // Handle responses from the async worker.
+            let mut empty = false;
+            while !empty {
+                match async_worker.try_recv() {
+                    Ok(response) => {
+                        let mut command_sequencer = command_sequencer.borrow_mut();
+                        self.handle_async_worker_response(&mut *command_sequencer, response);
+                    }
+                    Err(TryRecvError::Empty) => empty = true,
+                    Err(TryRecvError::Disconnected) => {
+                        empty = true;
+                    }
+                }
+            }
+
+            // Handle any received commands.
+            {
+                let mut command_sequencer = command_sequencer.borrow_mut();
+                while let Some(internal_cmd) = command_sequencer.next() {
+                    self.handle_internal_storage_command(&mut async_worker, internal_cmd);
+                }
```

Comment:
I think its _technically_ possible for this to starve the timely worker, as worker 0 could be producing messages faster than we can handle them; not sure if this is worth a comment saying we dont care about that edge case...